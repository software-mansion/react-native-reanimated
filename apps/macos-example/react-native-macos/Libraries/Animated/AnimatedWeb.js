/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @flow strict-local
 */

'use strict';

import AnimatedImplementation from './AnimatedImplementation';

export default {
  ...AnimatedImplementation,
  /* $FlowFixMe[incompatible-call] createAnimatedComponent expects to receive
   * types. Plain intrinsic components can't be typed like this */
  div: (AnimatedImplementation.createAnimatedComponent('div'): $FlowFixMe),
  /* $FlowFixMe[incompatible-call] createAnimatedComponent expects to receive
   * types. Plain intrinsic components can't be typed like this */
  span: (AnimatedImplementation.createAnimatedComponent('span'): $FlowFixMe),
  /* $FlowFixMe[incompatible-call] createAnimatedComponent expects to receive
   * types. Plain intrinsic components can't be typed like this */
  img: (AnimatedImplementation.createAnimatedComponent('img'): $FlowFixMe),
};
