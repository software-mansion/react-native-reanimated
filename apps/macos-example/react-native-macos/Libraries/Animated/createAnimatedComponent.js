/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 * @format
 */

import View from '../Components/View/View';
import useMergeRefs from '../Utilities/useMergeRefs';
import useAnimatedProps from './useAnimatedProps';
import * as React from 'react';

export type AnimatedComponentType<
  -Props: {+[string]: mixed, ...},
  +Instance = mixed,
> = React.AbstractComponent<
  $ObjMap<
    Props &
      $ReadOnly<{
        passthroughAnimatedPropExplicitValues?: React.ElementConfig<
          typeof View,
        >,
      }>,
    () => any,
  >,
  Instance,
>;

export default function createAnimatedComponent<TProps: {...}, TInstance>(
  Component: React.AbstractComponent<TProps, TInstance>,
): AnimatedComponentType<TProps, TInstance> {
  return React.forwardRef((props, forwardedRef) => {
    const [reducedProps, callbackRef] = useAnimatedProps<TProps, TInstance>(
      // $FlowFixMe[incompatible-call]
      props,
    );
    // $FlowFixMe[incompatible-call]
    const ref = useMergeRefs<TInstance | null>(callbackRef, forwardedRef);

    // Some components require explicit passthrough values for animation
    // to work properly. For example, if an animated component is
    // transformed and Pressable, onPress will not work after transform
    // without these passthrough values.
    // $FlowFixMe[prop-missing]
    const {passthroughAnimatedPropExplicitValues, style} = reducedProps;
    const {style: passthroughStyle, ...passthroughProps} =
      passthroughAnimatedPropExplicitValues ?? {};
    const mergedStyle = {...style, ...passthroughStyle};

    return (
      <Component
        {...reducedProps}
        {...passthroughProps}
        style={mergedStyle}
        ref={ref}
      />
    );
  });
}
