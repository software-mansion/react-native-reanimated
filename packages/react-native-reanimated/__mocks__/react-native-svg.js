const React = require('react');

const createComponent = function (name) {
  return class extends React.Component {
    static displayName = name;

    render() {
      return React.createElement(name, this.props, this.props.children);
    }

    setNativeProps(props) {
      const { style } = props;
      this.props = { ...this.props, ...style };
    }
  };
};

const Svg = createComponent('Svg');
const Path = createComponent('Path');
const Circle = createComponent('Circle');
const Rect = createComponent('Rect');
const G = createComponent('G');
const Line = createComponent('Line');

module.exports = { Svg, Path, Circle, Rect, G, Line };
