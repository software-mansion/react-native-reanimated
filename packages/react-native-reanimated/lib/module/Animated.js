'use strict';

export { ReanimatedFlatList as FlatList } from './component/FlatList';
export { AnimatedImage as Image } from './component/Image';
export { AnimatedScrollView as ScrollView } from './component/ScrollView';
export { AnimatedText as Text } from './component/Text';
export { AnimatedView as View } from './component/View';
export { addWhitelistedNativeProps, addWhitelistedUIProps } from './ConfigHelper';
export { createAnimatedComponent } from './createAnimatedComponent';
//# sourceMappingURL=Animated.js.map