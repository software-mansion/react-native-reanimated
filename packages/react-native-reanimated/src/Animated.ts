'use strict';
import type { Extrapolate as _Extrapolate } from './interpolateColor';
import type {
  SharedValue as _SharedValue,
  AnimatedTransform as _AnimatedTransform,
  AnimateStyle as _AnimateStyle,
  StylesOrDefault as _StylesOrDefault,
  EasingFunction as _EasingFunction,
} from './commonTypes';
import type { DerivedValue as _DerivedValue } from './hook/useDerivedValue';
import type {
  TransformStyleTypes as _TransformStyleTypes,
  Adaptable as _Adaptable,
  AdaptTransforms as _AdaptTransforms,
  AnimateProps as _AnimateProps,
} from './helperTypes';

import type { AnimatedScrollViewProps as _AnimatedScrollViewProps } from './component/ScrollView';
import type { FlatListPropsWithLayout as _FlatListPropsWithLayout } from './component/FlatList';

export { createAnimatedComponent } from './createAnimatedComponent';
export { AnimatedText as Text } from './component/Text';
export { AnimatedView as View } from './component/View';
export { AnimatedScrollView as ScrollView } from './component/ScrollView';
export { AnimatedImage as Image } from './component/Image';
export { ReanimatedFlatList as FlatList } from './component/FlatList';
export {
  addWhitelistedNativeProps,
  addWhitelistedUIProps,
} from './ConfigHelper';
/**
 * @deprecated Please import `Extrapolate` directly from
 *   `react-native-reanimated` instead of `Animated` namespace.
 */
export type Extrapolate = typeof _Extrapolate;
/**
 * @deprecated Please import `SharedValue` directly from
 *   `react-native-reanimated` instead of `Animated` namespace.
 */

export type SharedValue<T> = _SharedValue<T>;
/**
 * @deprecated Please import `DerivedValue` directly from
 *   `react-native-reanimated` instead of `Animated` namespace.
 */
export type DerivedValue<T> = _DerivedValue<T>;
/**
 * @deprecated Please import `Adaptable` directly from `react-native-reanimated`
 *   instead of `Animated` namespace.
 */
export type Adaptable<T> = _Adaptable<T>;
/**
 * @deprecated Please import `TransformStyleTypes` directly from
 *   `react-native-reanimated` instead of `Animated` namespace.
 */
export type TransformStyleTypes = _TransformStyleTypes;
/**
 * @deprecated Please import `AdaptTransforms` directly from
 *   `react-native-reanimated` instead of `Animated` namespace.
 */
export type AdaptTransforms<T> = _AdaptTransforms<T>;
/**
 * @deprecated Please import `AnimatedTransform` directly from
 *   `react-native-reanimated` instead of `Animated` namespace.
 */
export type AnimatedTransform = _AnimatedTransform;
/**
 * @deprecated Please import `AnimateStyle` directly from
 *   `react-native-reanimated` instead of `Animated` namespace.
 */
export type AnimateStyle<S> = _AnimateStyle<S>;
/**
 * @deprecated Please import `StylesOrDefault` directly from
 *   `react-native-reanimated` instead of `Animated` namespace.
 */
export type StylesOrDefault<S> = _StylesOrDefault<S>;
/**
 * @deprecated Please import `AnimateProps` directly from
 *   `react-native-reanimated` instead of `Animated` namespace.
 */
export type AnimateProps<P extends object> = _AnimateProps<P>;
/**
 * @deprecated Please import `EasingFunction` directly from
 *   `react-native-reanimated` instead of `Animated` namespace.
 */
export type EasingFunction = _EasingFunction;
/**
 * @deprecated Please import `AnimatedScrollViewProps` directly from
 *   `react-native-reanimated` instead of `Animated` namespace.
 */
export type AnimatedScrollViewProps = _AnimatedScrollViewProps;
/**
 * @deprecated Please import `FlatListPropsWithLayout` directly from
 *   `react-native-reanimated` instead of `Animated` namespace.
 */
export type FlatListPropsWithLayout<T> = _FlatListPropsWithLayout<T>;
