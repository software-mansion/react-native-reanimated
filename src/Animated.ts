import type { Extrapolate as _Extrapolate } from './reanimated2/interpolateColor';
import type { SharedValue as _SharedValue } from './reanimated2/commonTypes';
import type { DerivedValue as _DerivedValue } from './reanimated2/hook/useDerivedValue';
import type {
  TransformStyleTypes as _TransformStyleTypes,
  Adaptable as _Adaptable,
  AdaptTransforms as _AdaptTransforms,
  AnimatedTransform as _AnimatedTransform,
  AnimateStyle as _AnimateStyle,
  StylesOrDefault as _StylesOrDefault,
  AnimateProps as _AnimateProps,
} from './reanimated2/helperTypes';
import type { EasingFunction as _EasingFunction } from './reanimated2/Easing';
import {
  addWhitelistedNativeProps as _addWhitelistedNativeProps,
  addWhitelistedUIProps as _addWhitelistedUIProps,
} from './ConfigHelper';
import type { AnimatedScrollViewProps as _AnimatedScrollViewProps } from './reanimated2/component/ScrollView';
import type { FlatListPropsWithLayout as _FlatListPropsWithLayout } from './reanimated2/component/FlatList';
export { default as createAnimatedComponent } from './createAnimatedComponent';

export { AnimatedText as Text } from './reanimated2/component/Text';
export { AnimatedView as View } from './reanimated2/component/View';
export { AnimatedScrollView as ScrollView } from './reanimated2/component/ScrollView';
export { AnimatedImage as Image } from './reanimated2/component/Image';
export { ReanimatedFlatList as FlatList } from './reanimated2/component/FlatList';
/**
 * @deprecated Please import `Extrapolate` directly from `react-native-reanimated` instead of `Animated` namespace.
 */
export type Extrapolate = typeof _Extrapolate;
/**
 * @deprecated Please import `SharedValue` directly from `react-native-reanimated` instead of `Animated` namespace.
 */

export type SharedValue<T> = _SharedValue<T>;
/**
 * @deprecated Please import `DerivedValue` directly from `react-native-reanimated` instead of `Animated` namespace.
 */
export type DerivedValue<T> = _DerivedValue<T>;
/**
 * @deprecated Please import `Adaptable` directly from `react-native-reanimated` instead of `Animated` namespace.
 */
export type Adaptable<T> = _Adaptable<T>;
/**
 * @deprecated Please import `TransformStyleTypes` directly from `react-native-reanimated` instead of `Animated` namespace.
 * */
export type TransformStyleTypes = _TransformStyleTypes;
/**
 * @deprecated Please import `AdaptTransforms` directly from `react-native-reanimated` instead of `Animated` namespace.
 * */
export type AdaptTransforms<T> = _AdaptTransforms<T>;
/**
 * @deprecated Please import `AnimatedTransform` directly from `react-native-reanimated` instead of `Animated` namespace.
 */
export type AnimatedTransform = _AnimatedTransform;
/**
 * @deprecated Please import `AnimateStyle` directly from `react-native-reanimated` instead of `Animated` namespace.
 * */
export type AnimateStyle<S> = _AnimateStyle<S>;
/**
 * @deprecated Please import `StylesOrDefault` directly from `react-native-reanimated` instead of `Animated` namespace.
 * */
export type StylesOrDefault<S> = _StylesOrDefault<S>;
/**
 * @deprecated Please import `AnimateProps` directly from `react-native-reanimated` instead of `Animated` namespace.
 * */
export type AnimateProps<P extends object> = _AnimateProps<P>;
/**
 * @deprecated Please import `EasingFunction` directly from `react-native-reanimated` instead of `Animated` namespace.
 * */
export type EasingFunction = _EasingFunction;
/**
 * @deprecated Please import `addWhitelistedNativeProps` directly from `react-native-reanimated` instead of `Animated` namespace.
 * */
export const addWhitelistedNativeProps = _addWhitelistedNativeProps;
/**
 * @deprecated Please import `addWhitelistedUIProps` directly from `react-native-reanimated` instead of `Animated` namespace.
 * */
export const addWhitelistedUIProps = _addWhitelistedUIProps;
/**
 * @deprecated Please import `AnimatedScrollViewProps` directly from `react-native-reanimated` instead of `Animated` namespace.
 * */
export type AnimatedScrollViewProps = _AnimatedScrollViewProps;
/**
 * @deprecated Please import `FlatListPropsWithLayout` directly from `react-native-reanimated` instead of `Animated` namespace.
 * */
export type FlatListPropsWithLayout<T> = _FlatListPropsWithLayout<T>;
