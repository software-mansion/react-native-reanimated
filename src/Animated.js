import { LogBox } from 'react-native';
import createAnimatedComponent from './createAnimatedComponent';
import {
  addWhitelistedNativeProps,
  addWhitelistedUIProps,
} from './ConfigHelper';
import * as reanimated1 from './reanimated1';
import ReanimatedComponents from './reanimated2/component';

const Animated = {
  // components
  ...ReanimatedComponents,
  createAnimatedComponent,
  // configuration
  addWhitelistedNativeProps,
  addWhitelistedUIProps,
  // reanimated 1
  ...reanimated1,
};

export * from './reanimated2';
export * from './reanimated1';
export default Animated;

// I think we can ignore this message as long as Gesture Handler doesn't
// try to load the Reanimated module. I figured it should be here instead of
// RNGH because this prevents the message from being displayed for all
// versions of RNGH.
LogBox.ignoreLogs([
  'RCTBridge required dispatch_sync to load RNGestureHandlerModule. This may lead to deadlocks',
]);
