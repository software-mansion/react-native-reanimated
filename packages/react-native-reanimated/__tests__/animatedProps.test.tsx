import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';
import { Button, TextInput, View } from 'react-native';
import Animated, {
  useAnimatedProps,
  useSharedValue,
  withTiming,
} from 'react-native-reanimated';
import { Circle, Svg } from 'react-native-svg';

const animationDuration = 100;

const AnimatedCircle = Animated.createAnimatedComponent(Circle);
Animated.addWhitelistedNativeProps({ r: true });

const AnimatedTextInput = Animated.createAnimatedComponent(TextInput);
Animated.addWhitelistedNativeProps({ text: true });

export default function AnimatedComponent() {
  const r = useSharedValue(20);
  const width = useSharedValue(20);

  const handlePress = () => {
    r.value += 10;
    width.value += 10;
  };

  const animatedProps = useAnimatedProps(() => ({
    r: withTiming(r.value, { duration: animationDuration }),
  }));

  const textAnimatedProps = useAnimatedProps(() => {
    return {
      text: `Box width: ${width.value}`,
      defaultValue: `Box width: ${width.value}`,
    };
  });

  return (
    <View>
      <Svg>
        // SVG components strip our jest props and cannot be tested
        <AnimatedCircle
          cx="50%"
          cy="50%"
          fill="#b58df1"
          testID={'circle'}
          animatedProps={animatedProps}
        />
      </Svg>
      <AnimatedTextInput testID={'text'} animatedProps={textAnimatedProps} />
      <Button testID={'button'} onPress={handlePress} title="Click me" />
    </View>
  );
}

describe('animatedProps', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.runOnlyPendingTimers();
    jest.useRealTimers();
  });

  test('SVG component cannot be tested', () => {
    const { getByTestId } = render(<AnimatedComponent />);
    const circle = getByTestId('circle');

    expect(circle).toHaveAnimatedProps({});

    const rendered = render(<AnimatedComponent />).toJSON();
    expect(rendered).toMatchSnapshot();
  });

  test('Custom animated component', () => {
    const { getByTestId } = render(<AnimatedComponent />);
    const textInput = getByTestId('text');
    const button = getByTestId('button');

    expect(textInput).toHaveAnimatedProps({ text: 'Box width: 20' });

    fireEvent.press(button);
    jest.advanceTimersByTime(animationDuration);

    expect(textInput).toHaveAnimatedProps({ text: 'Box width: 30' });

    const rendered = render(<AnimatedComponent />).toJSON();
    expect(rendered).toMatchSnapshot();
  });
});
