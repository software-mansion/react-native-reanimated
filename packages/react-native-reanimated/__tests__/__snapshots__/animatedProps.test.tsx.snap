// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`animatedProps Custom animated component 1`] = `
<View>
  <RNSVGSvgView
    bbHeight="100%"
    bbWidth="100%"
    focusable={false}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "flex": 0,
          "height": "100%",
          "width": "100%",
        },
      ]
    }
  >
    <RNSVGGroup
      fill={
        {
          "payload": 4278190080,
          "type": 0,
        }
      }
    >
      // SVG components strip our jest props and cannot be tested
      <RNSVGCircle
        cx="50%"
        cy="50%"
        fill={
          {
            "payload": 4290088433,
            "type": 0,
          }
        }
        propList={
          [
            "fill",
          ]
        }
        r={20}
        testID="circle"
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <TextInput
    collapsable={false}
    defaultValue="Box width: 20"
    jestAnimatedProps={
      {
        "value": {
          "defaultValue": "Box width: 20",
          "text": "Box width: 20",
        },
      }
    }
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    testID="text"
    text="Box width: 20"
  />
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="button"
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Click me
      </Text>
    </View>
  </View>
</View>
`;

exports[`animatedProps SVG component cannot be tested 1`] = `
<View>
  <RNSVGSvgView
    bbHeight="100%"
    bbWidth="100%"
    focusable={false}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "flex": 0,
          "height": "100%",
          "width": "100%",
        },
      ]
    }
  >
    <RNSVGGroup
      fill={
        {
          "payload": 4278190080,
          "type": 0,
        }
      }
    >
      // SVG components strip our jest props and cannot be tested
      <RNSVGCircle
        cx="50%"
        cy="50%"
        fill={
          {
            "payload": 4290088433,
            "type": 0,
          }
        }
        propList={
          [
            "fill",
          ]
        }
        r={20}
        testID="circle"
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <TextInput
    collapsable={false}
    defaultValue="Box width: 20"
    jestAnimatedProps={
      {
        "value": {
          "defaultValue": "Box width: 20",
          "text": "Box width: 20",
        },
      }
    }
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    testID="text"
    text="Box width: 20"
  />
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="button"
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Click me
      </Text>
    </View>
  </View>
</View>
`;
